README.txt
(B-Cube) OWSLib
===============

Fork of the OWSLib project for improved GetCapabilities harmonization in 
support of the B-Cube service response harvesting project. 

This work is ongoing. Because of the focus on response parsing, changes
made to this fork, including but not limited to expanded version support,
may not adequately support data requests. So a WCS GetCapabilities 1.1.2
request should function but a GetCoverage request may not. 

Not all required changes have been implemented - this is unlikely to
install correctly at this time. 



Original Documentation
======================

OWSLib is a Python package for client programming with Open Geospatial
Consortium (OGC) web service (hence OWS) interface standards, and their
related content models.

Full documentation is available at http://geopython.github.io/OWSLib

OWSLib provides a common API for accessing service metadata and wrappers for
numerous OGC Web Service interfaces.

Dependencies
------------

OWSLib requires elementtree (standard in 2.5 as xml.etree) or lxml.

Usage
-----

Find out what a WMS has to offer. Service metadata::

    >>> from owslib.wms import WebMapService
    >>> wms = WebMapService('http://wms.jpl.nasa.gov/wms.cgi', version='1.1.1')
    >>> wms.identification.type
    'OGC:WMS'
    >>> wms.identification.version
    '1.1.1'
    >>> wms.identification.title
    'JPL Global Imagery Service'
    >>> wms.identification.abstract
    'WMS Server maintained by JPL, worldwide satellite imagery.'

Available layers::

    >>> list(wms.contents)
    ['us_landsat_wgs84', 'modis', 'global_mosaic_base', 'huemapped_srtm',
    'srtm_mag', 'daily_terra', 'us_ned', 'us_elevation', 'global_mosaic',
    'daily_terra_ndvi', 'daily_aqua_ndvi', 'daily_aqua_721', 'daily_planet',
    'BMNG', 'srtmplus', 'us_colordem', None, 'daily_aqua', 'worldwind_dem',
    'daily_terra_721']

Details of a layer::

    >>> wms['global_mosaic'].title
    'WMS Global Mosaic, pan sharpened'
    >>> wms['global_mosaic'].boundingBoxWGS84
    (-180.0, -60.0, 180.0, 84.0)
    >>> wms['global_mosaic'].crsOptions
    ['EPSG:4326', 'AUTO:42003']
    >>> wms['global_mosaic'].styles
    {'pseudo_bright': {'title': 'Pseudo-color image (Uses IR and Visual bands,
    542 mapping), gamma 1.5'}, 'pseudo': {'title': '(default) Pseudo-color
    image, pan sharpened (Uses IR and Visual bands, 542 mapping), gamma 1.5'},
    'visual': {'title': 'Real-color image, pan sharpened (Uses the visual
    bands, 321 mapping), gamma 1.5'}, 'pseudo_low': {'title': 'Pseudo-color
    image, pan sharpened (Uses IR and Visual bands, 542 mapping)'},
    'visual_low': {'title': 'Real-color image, pan sharpened (Uses the visual
    bands, 321 mapping)'}, 'visual_bright': {'title': 'Real-color image (Uses
    the visual bands, 321 mapping), gamma 1.5'}}

Available methods, their URLs, and available formats::

    >>> [op.name for op in wms.operations]
    ['GetTileService', 'GetCapabilities', 'GetMap']
    >>> wms.getOperationByName('GetMap').methods
    {'Get': {'url': 'http://wms.jpl.nasa.gov/wms.cgi?'}}
    >>> wms.getOperationByName('GetMap').formatOptions
    ['image/jpeg', 'image/png', 'image/geotiff', 'image/tiff']

That's everything needed to make a request for imagery::

    >>> img = wms.getmap(   layers=['global_mosaic'],
    ...                     styles=['visual_bright'],
    ...                     srs='EPSG:4326',
    ...                     bbox=(-112, 36, -106, 41),
    ...                     size=(300, 250),
    ...                     format='image/jpeg',
    ...                     transparent=True
    ...                     )
    >>> out = open('jpl_mosaic_visb.jpg', 'wb')
    >>> out.write(img.read())
    >>> out.close()

A very similar API exists for WebFeatureService. See
tests/wfs_MapServerWFSCapabilities.txt for details.

There is also support for Web Coverage Service (WCS), Catalogue
Service for the Web (CSW), Web Processing Service (WPS), and Web
Map Tile Service (WMTS). Some of those are beta quality.


Logging
-------
OWSLib logs messages to the 'owslib' named python logger. You may
configure your application to use the log messages like so:

    >>> import logging
    >>> owslib_log = logging.getLogger('owslib')
    >>> # Add formatting and handlers as needed
    >>> owslib_log.setLevel(logging.DEBUG)


Support
-------

http://lists.osgeo.org/mailman/listinfo/owslib-users
http://lists.osgeo.org/mailman/listinfo/owslib-devel


CHANGES.txt
Changes
=======

0.7 (unreleased, see bcube-mods branch) (2015-04-13)
---------------

- WMS 1.3.0 support
- WCS 1.x.x support
- Capabilities harmonization 
 - Capabilities Content harmonization
 - Capture of all bounding boxes and supported SRS/CRS options 


0.7 (2013-02-18)
---------------

- Support for SOS 1.0.0, SOS 2.0.0, SensorML (thanks @kwilcox)
- Support for TMS (thanks @cleder)
- numerous bug fixes

0.6 (2012-12-22)
----------------

- Support for WMTS (thanks @bradh)
- packaging support (thanks @kalxas) for:
 - openSUSE
 - Debian
- addition of owslib.__version__
- ISO support:
 - multiple gmd:identificationInfo elements
 - gmd:distributorInfo elements
- WMS
 - read additional Layer attributes (thanks @elemoine)
- numerous bug fixes

0.5 (2012-06-15)
----------------

- Support for the following parsers:
 - WPS 1.0.0
 - WFS 1.1.0
 - CRS handling
  - URNs
  - URIs
  - EPSG:xxxx style
- etree.py looks for lxml.etree first now
- catch WMS service exceptions on GetCapabilities
- CSW exceptions are now Pythonic

0.4 (2011-10-02)
----------------
- Support for the following parsers:
 - CSW 2.0.2
 - OWS Common 1.0.0, 1.1.0, 2.0.0
 - Filter Encoding 1.1.0
 - ISO 19115:2003
 - FGDC CSDGM
 - NASA DIF
 - Dublin Core
 - WFS 2.0
 - WCS 1.1
- New SCM/bug/mailing list infrastructure
- Sphinx documentation

0.3 (2008-05-08)
----------------
- WCS support.
- Support for basic authorization in WMS requests (#107).

0.2.1 (2007-08-06)
------------------
- Added support for Python 2.5.
- Fixed ticket #105: Don't depend on Content-length in the http headers for
  getfeature.

0.2.0 (2007-02-01)
------------------
- Change license to BSD.
- Added service contact metadata.


owslib/coverage/wcs110.py
# -*- coding: ISO-8859-15 -*-
# =============================================================================
# Copyright (c) 2004, 2006 Sean C. Gillies
# Copyright (c) 2007 STFC <http://www.stfc.ac.uk>
#
# Authors :
#          Dominic Lowe <d.lowe@rl.ac.uk>
#
# Contact email: d.lowe@rl.ac.uk
# =============================================================================

from __future__ import (absolute_import, division, print_function)

from .wcsBase import WCSBase, WCSCapabilitiesReader, ServiceException
from owslib.util import openURL, testXMLValue, xmltag_split
from urllib import urlencode
from owslib.crs import Crs

import logging
from owslib.util import log


def ns(tags):
    """go by the tag name, no namspaces for future/version-proofing
    """
    return '/'.join(['*[local-name()="%s"]' % t if t not in ['*', '..', '.'] else t
                     for t in tags.split('/') if t])


def find(elem, xpath):
    """assumes one element to return
    """
    return next(iter(elem.xpath(xpath)), None)


def findall(elem, xpath):
    return elem.xpath(xpath)


class WebCoverageService_1_1_x(WCSBase):
    """Abstraction for OGC Web Coverage Service (WCS), version 1.1.0
    Implements IWebCoverageService.
    """

    def __getitem__(self, name):
        ''' check contents dictionary to allow dict like access to service layers'''
        if name in self.__getattribute__('contents').keys():
            return self.__getattribute__('contents')[name]
        else:
            raise KeyError("No content named %s" % name)

    def __init__(self, url, xml=None, cookies=None, version='1.1.0'):
        self.version = version
        self.url = url
        self.cookies = cookies
        # initialize from saved capability document or access the server
        reader = WCSCapabilitiesReader(self.version)
        if xml:
            self._capabilities = reader.readString(xml)
        else:
            self._capabilities = reader.read(self.url)

        # check for exceptions
        se = find(self._capabilities, ns("Exception"))

        if se is not None:
            err_message = str(se.text).strip()
            raise ServiceException(err_message, xml)

        # build metadata objects:

        # serviceIdentification metadata
        elem = find(self._capabilities, ns('ServiceIdentification'))
        if elem is None:
            elem = find(self._capabilities, ns('ServiceIdentification'))
        self.identification = ServiceIdentification(elem)

        # serviceProvider
        elem = find(self._capabilities, ns('ServiceProvider'))
        self.provider = ServiceProvider(elem)

        # serviceOperations
        self.operations = []
        for elem in findall(self._capabilities, ns('OperationsMetadata/Operation')):
            self.operations.append(Operation(elem))

        # exceptions - ***********TO DO *************
            self.exceptions = [f.text for f
                               in findall(self._capabilities, ns('Capability/Exception/Format'))]

        # serviceContents: our assumption is that services use a top-level layer
        # as a metadata organizer, nothing more.
        self.contents = {}
        for elem in findall(self._capabilities, '/' + ns('*/Contents') + '//' + ns('CoverageSummary')):
            cm = ContentMetadata(elem, None, self)
            self.contents[cm.id] = cm
        # top = find(self._capabilities, ns('Contents/CoverageSummary'))
        # for elem in findall(self._capabilities, ns('Contents/CoverageSummary/CoverageSummary')):
        #     cm = ContentMetadata(elem, top, self)
        #     self.contents[cm.id] = cm

        # if self.contents == {}:
        #     # non-hierarchical.
        #     top = None
        #     for elem in findall(self._capabilities, ns('Contents/CoverageSummary')):
        #         cm = ContentMetadata(elem, top, self)
        #         # make the describeCoverage requests to
        #         # populate the supported formats/crs attributes
        #         self.contents[cm.id] = cm

    def items(self):
        '''supports dict-like items() access'''
        items = []
        for item in self.contents:
            items.append((item, self.contents[item]))
        return items


    #TO DO: Handle rest of the  WCS 1.1.0 keyword parameters e.g. GridCRS etc. 
    def getCoverage(self, identifier=None, bbox=None, time=None, format = None, store=False, rangesubset=None, gridbaseCRS=None, gridtype=None, gridCS=None, gridorigin=None, gridoffsets=None, method='Get',**kwargs):
        """Request and return a coverage from the WCS as a file-like object
        note: additional **kwargs helps with multi-version implementation
        core keyword arguments should be supported cross version
        example:
        cvg=wcs.getCoverageRequest(identifier=['TuMYrRQ4'], time=['2792-06-01T00:00:00.0'], bbox=(-112,36,-106,41),format='application/netcdf', store='true')

        is equivalent to:
        http://myhost/mywcs?SERVICE=WCS&REQUEST=GetCoverage&IDENTIFIER=TuMYrRQ4&VERSION=1.1.0&BOUNDINGBOX=-180,-90,180,90&TIMESEQUENCE=2792-06-01T00:00:00.0&FORMAT=application/netcdf
        
        if store = true, returns a coverages XML file
        if store = false, returns a multipart mime
        """
        if log.isEnabledFor(logging.DEBUG):
            log.debug('WCS 1.1.0 DEBUG: Parameters passed to GetCoverage: identifier=%s, bbox=%s, time=%s, format=%s, rangesubset=%s, gridbaseCRS=%s, gridtype=%s, gridCS=%s, gridorigin=%s, gridoffsets=%s, method=%s, other_arguments=%s'%(identifier, bbox, time, format, rangesubset, gridbaseCRS, gridtype, gridCS, gridorigin, gridoffsets, method, str(kwargs)))
        
        if method == 'Get':
            method='{http://www.opengis.net/wcs/1.1/ows}Get'
        try:
            base_url = next((m.get('url') for m in self.getOperationByName('GetCoverage').methods if m.get('type').lower() == method.lower()))
        except StopIteration:
            base_url = self.url


        #process kwargs
        request = {'version': self.version, 'request': 'GetCoverage', 'service':'WCS'}
        assert len(identifier) > 0
        request['identifier']=identifier
        #request['identifier'] = ','.join(identifier)
        if bbox:
            request['boundingbox']=','.join([repr(x) for x in bbox])
        if time:
            request['timesequence']=','.join(time)
        request['format']=format
        request['store']=store
        
        #rangesubset: untested - require a server implementation
        if rangesubset:
            request['RangeSubset']=rangesubset
        
        #GridCRS structure: untested - require a server implementation
        if gridbaseCRS:
            request['gridbaseCRS']=gridbaseCRS
        if gridtype:
            request['gridtype']=gridtype
        if gridCS:
            request['gridCS']=gridCS
        if gridorigin:
            request['gridorigin']=gridorigin
        if gridoffsets:
            request['gridoffsets']=gridoffsets
       
       #anything else e.g. vendor specific parameters must go through kwargs
        if kwargs:
            for kw in kwargs:
                request[kw]=kwargs[kw]
        
        #encode and request
        data = urlencode(request)
        
        u=openURL(base_url, data, method, self.cookies)
        return u

    def getOperationByName(self, name):
        """Return a named operation item."""
        for item in self.operations:
            if item.name == name:
                return item
        raise KeyError("No operation named %s" % name)


class Operation(object):
    """Abstraction for operation metadata
    Implements IOperationMetadata.
    """
    def __init__(self, elem):
        self.name = elem.get('name')
        # this is not valid
        # self.formatOptions = [f.text for f in findall(elem, ns('Parameter/AllowedValues/Value'))]
        self.formatOptions = []

        # for the constraints, to match the parameter values
        # and these are at the parent OperationMetadata level
        constraints = []
        for constraint in findall(elem, ns('../Constraint')):
            # let's just make that an or?
            cxp = ns('AllowedValues/Value') + ' | ' + ns('Value')
            constraints.append({
                constraint.attrib.get('name'): {
                    "values": [i.text for i in findall(constraint, cxp)]
                }
            })
        self.constraints = constraints

        methods = []
        for verb in findall(elem, ns('DCP/HTTP/*')):
            methods.append(
                {
                    "type": xmltag_split(verb.tag),
                    "url": verb.attrib['{http://www.w3.org/1999/xlink}href']
                }
            )
        self.methods = methods

        # for the parameters
        parameters = {}
        for parameter in findall(elem, ns('Parameter')):
            parameters[parameter.attrib['name']] = {'values': [i.text for i in findall(parameter,
                                                    ns('AllowedValues/Value'))]}
        self.parameters = parameters


class ServiceIdentification(object):
    """ Abstraction for ServiceIdentification Metadata
    implements IServiceIdentificationMetadata"""
    def __init__(self, elem):
        self.service = "WCS"
        # TODO: fix this
        self.version = "1.1.0"
        self.title = testXMLValue(find(elem, ns('Title')))
        self.abstract = testXMLValue(find(elem, ns('Abstract')))
        self.keywords = [f.text for f in findall(elem, ns('Keywords/Keyword'))]

        self.fees = testXMLValue(find(elem, ns('Fees')))

        self.accessconstraints = testXMLValue(find(elem, ns('AccessConstraints')))


class ServiceProvider(object):
    """ Abstraction for ServiceProvider metadata 
    implements IServiceProviderMetadata """
    def __init__(self,elem):
        name=elem.find('{http://www.opengis.net/ows}ProviderName')
        if name is not None:
            self.name=name.text
        else:
            self.name=None
        #self.contact=ServiceContact(elem.find('{http://www.opengis.net/ows}ServiceContact'))
        self.contact =ContactMetadata(elem)
        self.url=self.name # no obvious definitive place for url in wcs, repeat provider name?


class ContactMetadata(object):
    ''' implements IContactMetadata'''
    def __init__(self, elem):
        self.name = testXMLValue(find(elem, ns('ServiceContact/IndividualName')))

        self.organization = testXMLValue(find(elem, ns('ProviderName')))

        self.address = testXMLValue(find(elem,
                                    ns('ServiceContact/ContactInfo/Address/DeliveryPoint')))

        self.city = testXMLValue(find(elem, ns('ServiceContact/ContactInfo/Address/City')))

        self.region = testXMLValue(find(elem,
                                   ns('ServiceContact/ContactInfo/Address/AdministrativeArea')))

        self.postcode = testXMLValue(find(elem,
                                     ns('ServiceContact/ContactInfo/Address/PostalCode')))

        self.country = testXMLValue(find(elem, ns('ServiceContact/ContactInfo/Address/Country')))

        self.email = testXMLValue(find(elem,
                                  ns('ServiceContact/ContactInfo/Address/ElectronicMailAddress')))


class ContentMetadata(object):
    """Abstraction for WCS ContentMetadata
    Implements IContentMetadata
    """
    def __init__(self, elem, parent, service):
        """Initialize.

        CoverageSummary elements can be nested more than once and still be valid
        as XML. If the leaf does not have an identifier, the parent identifier
        should be used. SupportedCRS is the union of parent/child CRS values.

        There is nothing in the spec about inherited keyword sets. And no Description
        element?
        """
        # TODO - examine the parent for bounding box info.

        self._service = service
        self._elem = elem

        # find returns the first item so we should be good here
        self.id = testXMLValue(find(elem, '*[local-name()="Identifier"] | ../*[local-name()="Identifier"]'))
        self.abstract = testXMLValue(find(elem, '*[local-name()="Abstract"] | ../*[local-name()="Abstract"]'))
        self.title = testXMLValue(find(elem, '*[local-name()="Title"] | ../*[local-name()="Title"]'))

        # get the ancestors related to nested CoverageSummary elems and the local keywords
        tags = [elem.tag] + [e.tag for e in elem.iterancestors() if e.tag in ['CoverageSummary']]
        xpaths = ['/'.join(['..'] * i + ['*[local-name()="Keywords"]', '*[local-name()="Keyword"]'])
                  for i in xrange(len(tags))]
        self.keywords = [k.text for k in findall(elem, ' | '.join(xpaths))] if xpaths else []

        self.boundingBox = None  # needed for iContentMetadata harmonisation
        self.boundingBoxWGS84 = None
        b = find(elem, ns('WGS84BoundingBox'))
        if b is not None:
            lc = find(b, ns('LowerCorner')).text
            uc = find(b, ns('UpperCorner')).text
            self.boundingBoxWGS84 = (
                float(lc.split()[0]), float(lc.split()[1]),
                float(uc.split()[0]), float(uc.split()[1]),
            )

        # bboxes - other CRS
        self.boundingboxes = []
        for bbox in findall(elem, ns('BoundingBox')):
            if bbox is not None:
                try:
                    lc = find(b, ns('LowerCorner')).text
                    uc = find(b, ns('UpperCorner')).text
                    boundingBox = (
                        float(lc.split()[0]), float(lc.split()[1]),
                        float(uc.split()[0]), float(uc.split()[1]),
                        b.attrib['crs']
                    )
                    self.boundingboxes.append(boundingBox)
                except:
                    pass

        # others not used but needed for iContentMetadata harmonisation
        self.styles = None
        self.crsOptions = None

        # SupportedCRS
        self.supportedCRS = []
        for crs in elem.findall('{http://www.opengis.net/wcs/1.1}SupportedCRS'):
            self.supportedCRS.append(Crs(crs.text))

        # SupportedFormats
        self.supportedFormats = []
        for format in findall(elem, ns('SupportedFormat')):
            self.supportedFormats.append(format.text)

    # grid is either a gml:Grid or a gml:RectifiedGrid if supplied as part of the DescribeCoverage response.
    def _getGrid(self):
        grid=None
        #TODO- convert this to 1.1 from 1.0
        #if not hasattr(self, 'descCov'):
                #self.descCov=self._service.getDescribeCoverage(self.id)
        #gridelem= self.descCov.find(ns('CoverageOffering/')+ns('domainSet/')+ns('spatialDomain/')+'{http://www.opengis.net/gml}RectifiedGrid')
        #if gridelem is not None:
            #grid=RectifiedGrid(gridelem)
        #else:
            #gridelem=self.descCov.find(ns('CoverageOffering/')+ns('domainSet/')+ns('spatialDomain/')+'{http://www.opengis.net/gml}Grid')
            #grid=Grid(gridelem)
        return grid
    grid = property(_getGrid, None)

    # time limits/postions require a describeCoverage request therefore only resolve when requested
    def _getTimeLimits(self):
        timelimits = []
        for elem in findall(self._service.getDescribeCoverage(self.id), ns('CoverageDescription/Domain/TemporalDomain/TimePeriod')):
            subelems = elem.getchildren()
            timelimits = [subelems[0].text, subelems[1].text]
        return timelimits
    timelimits = property(_getTimeLimits, None)

    # TODO timepositions property
    def _getTimePositions(self):
        return []
    timepositions = property(_getTimePositions, None)
